### ISSUE TITLE
zlib.error: Error -3 while decompressing data: invalid stored block lengths

### ISSUE THREAD

**Issue by brechtm** [superatomic/tldr-man#3]
```shell
$ tldr --update
Updating tldr-pages cache...
pages       / windows  [####################################]  100%
pages       / osx      [####################################]  100%
pages       / common   [#####-------------------------------]   14%  00:00:06
Traceback (most recent call last):
  File "/apps/hosted/tldr-man/.venv/lib/python3.10/site-packages/tldr_man/main.py", line 59, in wrapper
    func(ctx, param, value)
  File "/apps/hosted/tldr-man/.venv/lib/python3.10/site-packages/tldr_man/main.py", line 80, in subcommand_update
    pages.update_cache()
  File "/apps/hosted/tldr-man/.venv/lib/python3.10/site-packages/tldr_man/pages.py", line 165, in update_cache
    for filename, manpage in fut:
  File "/apps/hosted/tldr-man/.venv/lib/python3.10/site-packages/click/_termui_impl.py", line 328, in generator
    for rv in self.iter:
  File "/apps/hosted/spack/opt/spack/linux-rhel7-haswell/gcc-4.8.5/python-3.10.1-ek76ski4ctzhij7smku7sh3bmupopkou/lib/python3.10/concurrent/futures/_base.py", line 608, in result_iterator
    yield fs.pop().result()
  File "/apps/hosted/spack/opt/spack/linux-rhel7-haswell/gcc-4.8.5/python-3.10.1-ek76ski4ctzhij7smku7sh3bmupopkou/lib/python3.10/concurrent/futures/_base.py", line 438, in result
    return self.__get_result()
  File "/apps/hosted/spack/opt/spack/linux-rhel7-haswell/gcc-4.8.5/python-3.10.1-ek76ski4ctzhij7smku7sh3bmupopkou/lib/python3.10/concurrent/futures/_base.py", line 390, in __get_result
    raise self._exception
  File "/apps/hosted/spack/opt/spack/linux-rhel7-haswell/gcc-4.8.5/python-3.10.1-ek76ski4ctzhij7smku7sh3bmupopkou/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/apps/hosted/tldr-man/.venv/lib/python3.10/site-packages/tldr_man/pages.py", line 147, in to_manpage
    rendered_manpage = render_manpage(tldr_page.read_text())
  File "/apps/hosted/spack/opt/spack/linux-rhel7-haswell/gcc-4.8.5/python-3.10.1-ek76ski4ctzhij7smku7sh3bmupopkou/lib/python3.10/zipfile.py", line 2352, in read_text
    return strm.read()
  File "/apps/hosted/spack/opt/spack/linux-rhel7-haswell/gcc-4.8.5/python-3.10.1-ek76ski4ctzhij7smku7sh3bmupopkou/lib/python3.10/zipfile.py", line 909, in read
    buf += self._read1(self.MAX_N)
  File "/apps/hosted/spack/opt/spack/linux-rhel7-haswell/gcc-4.8.5/python-3.10.1-ek76ski4ctzhij7smku7sh3bmupopkou/lib/python3.10/zipfile.py", line 999, in _read1
    data = self._decompressor.decompress(data, n)
zlib.error: Error -3 while decompressing data: invalid stored block lengths
```

Running this again, this happens at different points in the process. For example, on another run, it crashed at converting the linux man pages. And with another error message:
```shell
$ tldr --update
Updating tldr-pages cache...
pages       / windows  [####################################]  100%
pages       / osx      [####################################]  100%
pages       / common   [####################################]  100%
pages       / sunos    [####################################]  100%
pages       / android  [####################################]  100%
pages       / linux    [################################----]   91%  00:00:00
Traceback (most recent call last):
  File "/apps/hosted/tldr-man/.venv/lib/python3.10/site-packages/tldr_man/main.py", line 59, in wrapper
    func(ctx, param, value)
  File "/apps/hosted/tldr-man/.venv/lib/python3.10/site-packages/tldr_man/main.py", line 80, in subcommand_update
    pages.update_cache()
  File "/apps/hosted/tldr-man/.venv/lib/python3.10/site-packages/tldr_man/pages.py", line 165, in update_cache
    for filename, manpage in fut:
  File "/apps/hosted/tldr-man/.venv/lib/python3.10/site-packages/click/_termui_impl.py", line 328, in generator
    for rv in self.iter:
  File "/apps/hosted/spack/opt/spack/linux-rhel7-haswell/gcc-4.8.5/python-3.10.1-ek76ski4ctzhij7smku7sh3bmupopkou/lib/python3.10/concurrent/futures/_base.py", line 608, in result_iterator
    yield fs.pop().result()
  File "/apps/hosted/spack/opt/spack/linux-rhel7-haswell/gcc-4.8.5/python-3.10.1-ek76ski4ctzhij7smku7sh3bmupopkou/lib/python3.10/concurrent/futures/_base.py", line 438, in result
    return self.__get_result()
  File "/apps/hosted/spack/opt/spack/linux-rhel7-haswell/gcc-4.8.5/python-3.10.1-ek76ski4ctzhij7smku7sh3bmupopkou/lib/python3.10/concurrent/futures/_base.py", line 390, in __get_result
    raise self._exception
  File "/apps/hosted/spack/opt/spack/linux-rhel7-haswell/gcc-4.8.5/python-3.10.1-ek76ski4ctzhij7smku7sh3bmupopkou/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/apps/hosted/tldr-man/.venv/lib/python3.10/site-packages/tldr_man/pages.py", line 147, in to_manpage
    rendered_manpage = render_manpage(tldr_page.read_text())
  File "/apps/hosted/spack/opt/spack/linux-rhel7-haswell/gcc-4.8.5/python-3.10.1-ek76ski4ctzhij7smku7sh3bmupopkou/lib/python3.10/zipfile.py", line 2352, in read_text
    return strm.read()
  File "/apps/hosted/spack/opt/spack/linux-rhel7-haswell/gcc-4.8.5/python-3.10.1-ek76ski4ctzhij7smku7sh3bmupopkou/lib/python3.10/zipfile.py", line 909, in read
    buf += self._read1(self.MAX_N)
  File "/apps/hosted/spack/opt/spack/linux-rhel7-haswell/gcc-4.8.5/python-3.10.1-ek76ski4ctzhij7smku7sh3bmupopkou/lib/python3.10/zipfile.py", line 999, in _read1
    data = self._decompressor.decompress(data, n)
zlib.error: Error -3 while decompressing data: invalid distance too far back
```

**Comment by brechtm** [comment]
I tried with Python 3.10.8, which does work! I don't know whether it's an issue with Python 3.10.1 or with my particular installation.

**Comment by superatomic** [comment]
Thank you for bringing this issue to my attention!

I was able to reproduce a similar error in `tldr --update` using Python 3.10.1 and Python 3.10.3, resulting in the following error messages:
- `zipfile.BadZipFile: Bad magic number for file header`
- `zlib.error: Error -3 while decompressing data: invalid distance too far back`
- `zlib.error: Error -3 while decompressing data: invalid code lengths set`

The cause of this problem seems to be that [`read_text()` method of `zipfile.Path`][1] is not thread-safe. This is a known bug in the Python Standard Library, which was reported in https://github.com/python/cpython/issues/86535 ([bpo-42369][2]) and was fixed in Python version 3.10.4.

Since **tldr-man-client v1.1.0** introduced the use of multiple threads for `tldr --update` in #2, this issue has only been present in that version, and only when using Python versions 3.10.0, 3.10.1, 3.10.2, and 3.10.3.

Since this is a bug with certain Python versions, this issue can be resolved by updating [`pyproject.toml`][3] to depend on a `python` version of `^3.10.4` instead of just `^3.10`.



[1]: https://docs.python.org/3.10/library/zipfile.html#zipfile.Path.read_text
[2]: https://bugs.python.org/issue42369
[3]: https://github.com/superatomic/tldr-man-client/blob/85bf23cb43130cfe23a994bc2f35be183df2191d/pyproject.toml#L34

**Comment by superatomic** [comment]
Thanks again for reporting this bug! The issue should be fixed in **tldr-man-client v1.1.1** ([PyPI][1], [GitHub][2]).

Please let me know if the issue persists.

[1]: https://pypi.org/project/tldr-man/1.1.1/
[2]: https://github.com/superatomic/tldr-man-client/releases/tag/v1.1.1
